### PR TITLE
Symintify pad ops

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87304
* #87291
* #87231
* __->__ #87046

Following comments below, we need to add support for `std::negate`/`std::min`/`std::max`/`operator-` for SymInt

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -4859,15 +4859,15 @@ Tensor sparse_constructor_values_backward(
...
 // in pads])
-Tensor constant_pad_nd_backward(const Tensor& grad, IntArrayRef pad) {
+Tensor constant_pad_nd_backward(const Tensor& grad, c10::SymIntArrayRef pad) {
   auto negated_pad = pad.vec();
   // NOLINTNEXTLINE(modernize-use-transparent-functors)
   std::transform(
       negated_pad.cbegin(),
       negated_pad.cend(),
       negated_pad.begin(),
-      std::negate<int64_t>());
```
Comment (ezyang) [inline torch/csrc/autograd/FunctionsManual.cpp:4869]: We should add a std::negate specialization for SymInt. We should also make operator- work. With the branch, there isn't any particular reason to rush this

File: aten/src/ATen/native/PadNd.cpp [hunk 2]
```diff
@@ -137,63 +137,63 @@ Tensor _pad_circular(const Tensor &self, IntArrayRef padding) {
...
+                           out_shape[dim] - pad_l - pad_r.max(zero),
+                           out_shape[dim] - pad_r.max(zero));
       out_slice.copy_(in_slice);
     }
 
     if (pad_r > 0) {
-      out_slice = out.slice(dim, out_shape[dim] - pad_r, out_shape[dim]);
-      in_slice = out.slice(dim, std::max(pad_l, zero), std::max(pad_l, zero) + pad_r);
+      out_slice = out.slice_symint(dim, out_shape[dim] - pad_r, out_shape[dim]);
+      in_slice = out.slice_symint(dim, pad_l.max(zero), pad_l.max(zero) + pad_r);
```
Comment (ezyang) [inline aten/src/ATen/native/PadNd.cpp:149]: We should have a SymInt specialization for std::max

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/PadNd.cpp (hunk 2 above)
Comment (bdhirsh) [reply aten/src/ATen/native/PadNd.cpp:149]: I added one last week but ended up removing it.

From the [docs](https://en.cppreference.com/w/cpp/algorithm/max), it looks like `std::max/min()` already have a template that just calls into `operator<` and `operator>`. So adding a special handling for min/max on symints in C++ seemed redundant to me, but lmk if you think that's wrong. We still have min/max support in python.

Comment (ezyang) [reply aten/src/ATen/native/PadNd.cpp:149]: ok, I am fine with this resolution

**Comment by albanD** [comment]
@ezyang this is ready for a final review

**Pull request closed without merging**